README.md
# quicksilver
![Quicksilver Logo](./logo.svg)

[![Crates.io](https://img.shields.io/crates/v/quicksilver.svg)](https://crates.io/crates/quicksilver)
[![Docs Status](https://docs.rs/quicksilver/badge.svg)](https://docs.rs/quicksilver)
[![dependency status](https://deps.rs/repo/github/ryanisaacg/quicksilver/status.svg)](https://deps.rs/repo/github/ryanisaacg/quicksilver)

A simple 2D game framework written in pure Rust, for both the Web and Desktop

## Maintenance Status

I've [posted an update on my website](https://ryanisaacg.com/posts/quicksilver-goodbye) about Quicksilver.
To keep a long story short: **Quicksilver is no longer actively developed.**
For now I will continue to triage bugs and pull requests and (maybe) fix small bugs.


## Alpha Notice

This version of Quicksilver is currently working its way through alpha! There is still work to do
on the API and on bugfixes, as well as waiting on an upstream library for audio support.
Please feel free to use this version and **provide feedback!** If you run into bugs or want to
give feedback on API decisions, please open an issue.

## A quick example

Create a rust project and add this line to your `Cargo.toml` file under `[dependencies]`:
```text
quicksilver = "0.4"
```
Then replace `src/main.rs` with the following (the contents of quicksilver's
`examples/01_square.rs`):

```rust
// Example 1: The Square
// Open a window, and draw a colored square in it
use quicksilver::{
    geom::{Rectangle, Vector},
    graphics::Color,
    run, Graphics, Input, Result, Settings, Window,
};

fn main() {
    run(
        Settings {
            title: "Square Example",
            ..Settings::default()
        },
        app,
    );
}

async fn app(window: Window, mut gfx: Graphics, mut input: Input) -> Result<()> {
    // Clear the screen to a blank, white color
    gfx.clear(Color::WHITE);
    // Paint a blue square with a red outline in the center of our screen
    // It should have a top-left of (350, 100) and a size of (150, 100)
    let rect = Rectangle::new(Vector::new(350.0, 100.0), Vector::new(100.0, 100.0));
    gfx.fill_rect(&rect, Color::BLUE);
    gfx.stroke_rect(&rect, Color::RED);
    // Send the data to be drawn
    gfx.present(&window)?;
    loop {
        while let Some(_) = input.next_event().await {}
    }
}
```

## Learning Quicksilver

A good way to get started with Quicksilver is to
[read and run the examples](https://github.com/ryanisaacg/quicksilver/tree/master/examples)
which also serve as tutorials. If you have any questions, feel free to open an issue or ask for
help in the [Rust Community Discord](https://discord.gg/aVESxV8) from other
Quicksilver users and developers.

## Made with Quicksilver

### Version 0.4

- Libraries
    - [Lenscas](https://github.com/lenscas): [Silver_Animation](https://crates.io/crates/silver_animation) - An animation system
    - [Lenscas](https://github.com/lenscas): [Mergui](https://crates.io/crates/mergui) - A simple GUI system
    - [johnpmayer](https://github.com/johnpmayer): [quicksilver-utils-async](https://crates.io/crates/quicksilver-utils-async) - Tasks, timers, and net code
    - [johnpmayer](https://github.com/johnpmayer): [quicksilver-utils-ecs](https://crates.io/crates/quicksilver-utils-ecs) - Entity Component System integrations
- Games
    - [alec-deason](https://github.com/alec-deason): [Pixel Imperfect](https://ntoheuns.itch.io/pixel-imperfect)

### Version 0.3

- Documentation / Tutorials
    - [tomassedovic](https://github.com/tomassedovic): [quicksilver-roguelike](https://github.com/tomassedovic/quicksilver-roguelike)
- Games
    - [WushuWorks](https://github.com/WushuWorks): [I am the Elder God](https://wushuworks.github.io/I-am-the-Elder-God/)
    - [codec-abc](https://github.com/codec-abc): [RustyVolley](https://github.com/RustyVolley/RustyVolleySrc)
    - [rickyhan](https://github.com/rickyhan): [Kingston Crabfight Simulator](https://github.com/rickyhan/crabs)
    - [robotcaleb](https://github.com/robotcaleb): [Replay](https://robotcaleb.github.io/Replay/)
    - [rsribeiro](https://github.com/rsribeiro/): [Evil Alligator](https://rsribeiro.github.io/website/)
    - [nycex](https://gitlab.com/nycex): [Axosnake](https://gitlab.com/nycex/axosnake)
    - [Leinnan](https://github.com/Leinnan): [Slavic Castles](https://github.com/Leinnan/slavic_castles)
    - [Lenscas](https://github.com/lenscas): [Arena keeper](https://github.com/lenscas/arena_keeper_quick)
- Libraries
    - [Lenscas](https://github.com/lenscas): [Mergui](https://crates.io/crates/mergui) - A simple GUI system

Want to add your project? Feel free to open an issue or PR!

## Building and Deploying a Quicksilver application

Quicksilver should always compile and run on the latest stable version of Rust, for both web and
desktop.

Make sure to put all your assets in a top-level folder of your crate called `static/`. *All*
Quicksilver file loading-APIs will expect paths that originate in the static folder, so
`static/image.png` should be referenced as `image.png`.

### Linux dependencies

On Windows and Mac, all you'll need to build Quicksilver is a recent stable version of `rustc`
and `cargo`. A few of Quicksilver's dependencies require Linux packages to build, namely
`libudev`, `zlib`, and `alsa`. To install these on Ubuntu or Debian, run the command
`sudo apt install libudev-dev zlib1g-dev alsa libasound2-dev`.

### Deploying for desktop

If you're deploying for desktop platforms, build in release mode (`cargo build --release`)
and copy the executable file produced (found at "target/release/") and any assets you used
(image files, etc.) and create an archive (on Windows a zip file, on Unix a tar file). You
should be able to distribute this archive with no problems; if there are any, please open an
issue.

### Deploying for the web

If you're deploying for the web, first make sure you've
[installed the cargo web tool](https://github.com/koute/cargo-web). Then use `cargo web deploy`
to build your application for distribution (located at `target/deploy`).

If you want to test your application locally, use `cargo web start --features quicksilver/stdweb` and open your
favorite browser to the port it provides.

#### wasm-bindgen support

Quicksilver has recently gained experimental support for `wasm-bindgen`, under the `web-sys`
feature. The workflow is not currently documented here, but it should be the same as using any other
library with `wasm-bindgen`.

## Optional Features

Quicksilver by default tries to provide all features a 2D application may need, but not all
applications need these features.

The optional features available are:
- easy logging (via [log](https://github.com/rust-lang/log),
[simple_logger](https://github.com/borntyping/rust-simple_logger), and
[web_logger](https://github.com/yewstack/web_logger))
- gamepad event generation (via [gilrs](https://gitlab.com/gilrs-project/gilrs))
- saving (via [gestalt](https://github.com/ryanisaacg/gestalt))
- font rendering (via [elefont](https://github.com/ryanisaacg/elefont)) and TTF parsing (via [rusttype](https://gitlab.redox-os.org/redox-os/rusttype))

Each are enabled by default, but you can
[specify which features](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#choosing-features)
you actually want to use.

## Supported Platforms

The engine is supported on Windows, macOS, Linux, and the web via WebAssembly.

Mobile support would be a future possibility, but likely only through external contributions.



CHANGES.md
# Changelog

- #\[inline] small functions

## v0.4.0
- Derive Debug and Clone on `Timer`
- Apply the inverse of `set_view` in `screen_to_camera`
- Updated status to `passively-maintained`

## v0.4.0-alpha0.7
- Remove stray calls to `dbg!`

## v0.4.0-alpha0.6
- Do not crash if logging is already initalized
- Add optional dependency on serde, when enabled it adds the Serialize and Deserialize traits to Circle, Vector and Rectangle
- Fix a bug where text extents were improperly reported
- [BREAKING] Remove `set_viewport`, `fit_to_surface`, and `fit_to_window`; the viewport is now set automatically
- [BREAKING] Remove `flush(Option<Surface>)` in favor of `flush_surface` and `flush_window`
- [BREAKING] Remove `Graphics::set_projection` in favor of `Graphics::set_view`
- [BREAKING] Remove `transform_for_size` and `projection` from `ResizeHandler`, in favor of `Graphics::set_resize_handler`
- Add the `set_camera_size` to `Graphics` to manage the new virtual camera abstraction
- Add `screen_to_camera` to `Graphics` for mapping the mouse position to the camera space
- Fix images failing to draw when their destructors run before they are flushed

## v0.4.0-alpha0.5
- Fix `Timer::remaining` returning the time until next tick, instead of returning how late the tick is.
- Add methods to `Timer`: `late_by`, `period`, and `elapsed`
- [BREAKING] Remove `Scalar` and replace it with `f32`
- [BREAKNNG] Remove all uses of `impl Into<Vector>` in the public API
- Deprecate `Line`, `Triangle`, and a handful of `Shape` methods
- [BREAKING] Automatically re-set the viewport to the size of the window when `present` is called
- Actually add `ResizeHandler` to help with maintaing the same aspect ratio when the screen size changes

## v0.4.0-alpha0.4
- Fix compile issues with font-related features
- [BREAKING] Replace 'lifecycle' module with 'input' module:
    - [BREAKING] Rename `EventStream` to `Input`
    - Integrate the input state cache directly into `Input`
    - [BREAKING] The `blinds::Window` struct and the `Event` enums are now wrapped with methods that use `quicksilver::geom::Vector` instead of `mint::Vector2`
- Implement `From` instead of `Into` for some types

## v0.4.0-alpha0.3
- Update `golem` to `v0.1.1` to fix non-power-of-2 textures
- `impl std::iter::Sum for geom::Vector`
- Implement `std::ops::MulAssign`, `std::ops::AddAssign`, and `std::ops::SubAssign` for `Transform`s.
- Addition and subtraction of Tranforms supported to help with easing function calculations
- Added an example with loading progress bar
- Add `Timer` struct to help timing draw calls and a consistent update cycle
- Add `exhaust` and `reset` function to timer so they can be used for more than just an update cycle
- `lifecycle::run` can now accept any kind of Error.
- Add `into_raw_context` on Graphics, to allow lower-level graphics programming
- Add font support!
  - `VectorFont` allows you to load TTF files via rusttype
  - `FontRenderer` allows you to draw glyphs to the screen

## v0.4.0-alpha0.2
- Fix the "easy-log" feature
- [BREAKING] `fill_circle` now takes a `&Circle` instead of a separate center and radius

## 0.4.0-alpha0.1
- Fix the default blend mode: it should mix colors using their alphas
- [Breaking] Remove support for image types other than jpeg and png
- [Breaking] Add a parameter to `Graphics::flush` which determines what Surface to render to
- Add the Surface API, for rendering to textures
- Add functions to set texture parameters and to set texture data
- Add the ability to set the blend pipeline
- Fix over-estimation of GPU buffer sizes
- Add functions on Graphics to set the viewport (`set_viewport`, `fit_to_surfcae`, `fit_to_window`)
- Add `ResizeHandler` to help with maintaing the same aspect ratio when the screen size changes
- Re-export `blinds`, `golem`, `mint`, and `log`
- [Breaking] Update to blinds 0.1.0:
    - [Breaking] Reworked the events API for forwards-compatibility
    - Set the tab title on web
    - Add an optional module to cache event states

## 0.4.0-alpha0
The API change is *very breaking*. It can be considered nearly a full re-write of Quicksilver.

- Added the new async API, via `blinds`
- Added experimental support for `web-sys` behind the feature named `web-sys`
- [Breaking] Moved `stdweb` support behind the feature named `stdweb`
- [Breaking] Removed the following APIs/integrations, pending re-works (to be added before 0.4):
    - BlendMode
    - Font
    - ImageScaleStrategy
    - ResizeStrategy
    - Sound
    - Surface
    - `lyon` integration
- [Breaking] Removed the following APIs/integrations, (possibly to be added before 0.4);
    - Keyboard
    - Mouse
- [Breaking] Removed the following APIs/integrations permanently:
    - Asset
    - Animation
    - Atlas
    - State
    - combinators module
    - `immi` integration
    - `nalgebra` integration (replaced with `mint` integration)

## 0.3.21
- Fix gamepad buttons mistakenly marked as `Released` instead of `NotPressed`

## 0.3.20
- Fix gamepad Select/Start events misfiring when A is pressed

## 0.3.19
- Update dependencies
- Bump WebGL requirement from 2.0 down to 1.0

## 0.3.18
- Fix the circle fix (was mistakenly applied to triangles)

## 0.3.17
- Add `transform_bounding_box` to `Shape`
- Fix a bug with circles being textured incorrectly

## 0.3.16
- Fix an issue creating non-RGBA images on web
- Deprecate Animation and Immi integration, pending removal in a future release

## 0.3.15
- Fix an issue with the viewport in `Surface::render_to()`
- Fix a GL error in `Image::new_null()` for the WebGL backend
- Fix a JavaScript error for WebGL backend in Firefox
- Unbind newly created FBOs and textures to prevent GL errors and unwanted behaviours

## 0.3.14
- Documentation fixes in src/lib.rs

## 0.3.13
- Allow any type that is `Into<Background>` to be passed into `draw` and `draw_ex`
- Update docs for `save` and `load`

## 0.3.12
- Updated glutin to 0.21
- Expand the Event::Typed to include punctuation

## 0.3.11

src/lib.rs
//! # quicksilver
//! ![Quicksilver Logo](https://raw.github.com/ryanisaacg/quicksilver/master/logo.svg?sanitize=true)
//!
//! [![Crates.io](https://img.shields.io/crates/v/quicksilver.svg)](https://crates.io/crates/quicksilver)
//! [![Docs Status](https://docs.rs/quicksilver/badge.svg)](https://docs.rs/quicksilver)
//! [![dependency status](https://deps.rs/repo/github/ryanisaacg/quicksilver/status.svg)](https://deps.rs/repo/github/ryanisaacg/quicksilver)
//!
//! A simple 2D game framework written in pure Rust, for both the Web and Desktop
//!
//! ## Maintenance Status
//!
//! I've [posted an update on my website](https://ryanisaacg.com/posts/quicksilver-goodbye) about Quicksilver.
//! To keep a long story short: **Quicksilver is no longer actively developed.**
//! For now I will continue to triage bugs and pull requests and (maybe) fix small bugs.
//!
//! ## Alpha Notice
//!
//! This version of Quicksilver is currently working its way through alpha! There is still work to do
//! on the API and on bugfixes, as well as waiting on an upstream library for audio support.
//! Please feel free to use this version and **provide feedback!** If you run into bugs or want to
//! give feedback on API decisions, please open an issue.
//!
//! ## A quick example
//!
//! Create a rust project and add this line to your `Cargo.toml` file under `[dependencies]`:
//! ```text
//!     quicksilver = "0.4"
//! ```
//! Then replace `src/main.rs` with the following (the contents of quicksilver's
//! `examples/01_square.rs`):
//!
//! ```no_run
//! // Example 1: The Square
//! // Open a window, and draw a colored square in it
//! use quicksilver::{
//!     geom::{Rectangle, Vector},
//!     graphics::{Color, Graphics},
//!     Input, Window, Result, Settings, run,
//! };
//!
//! fn main() {
//!     run(
//!         Settings {
//!             title: "Square Example",
//!             ..Settings::default()
//!         },
//!         app,
//!     );
//! }
//!
//! async fn app(window: Window, mut gfx: Graphics, mut input: Input) -> Result<()> {
//!     // Clear the screen to a blank, white color
//!     gfx.clear(Color::WHITE);
//!     // Paint a blue square with a red outline in the center of our screen
//!     // It should have a top-left of (350, 100) and a size of (150, 100)
//!     let rect = Rectangle::new(Vector::new(350.0, 100.0), Vector::new(100.0, 100.0));
//!     gfx.fill_rect(&rect, Color::BLUE);
//!     gfx.stroke_rect(&rect, Color::RED);
//!     // Send the data to be drawn
//!     gfx.present(&window)?;
//!     loop {
//!         while let Some(_) = input.next_event().await {}
//!     }
//! }
//! ```
//!
//! Run this with `cargo run` or, if you have the wasm32 toolchain installed, you can build for the
//! web (instructions below).
//!
//! ## Learning Quicksilver
//!
//! A good way to get started with Quicksilver is to
//! [read and run the examples](https://github.com/ryanisaacg/quicksilver/tree/master/examples)
//! which also serve as tutorials. If you have any questions, feel free to open an issue or ask for
//! help in the [Rust Community Discord](https://discord.gg/aVESxV8) from other Quicksilver users
//! and developers.
//!
//! ## Building and Deploying a Quicksilver application
//!
//! Quicksilver should always compile and run on the latest stable version of Rust, for both web and
//! desktop.
//!
//! Make sure to put all your assets in a top-level folder of your crate called `static/`. *All*
//! Quicksilver file loading-APIs will expect paths that originate in the static folder, so
//! `static/image.png` should be referenced as `image.png`.
//!
//! ### Linux dependencies
//!
//! On Windows and Mac, all you'll need to build Quicksilver is a recent stable version of `rustc`
//! and `cargo`. A few of Quicksilver's dependencies require Linux packages to build, namely
//! `libudev`, `zlib`, and `alsa`. To install these on Ubuntu or Debian, run the command
//! `sudo apt install libudev-dev zlib1g-dev alsa libasound2-dev`.
//!
//! ### Deploying for desktop
//!
//! If you're deploying for desktop platforms, build in release mode (`cargo build --release`)
//! and copy the executable file produced (found at "target/release/") and any assets you used
//! (image files, etc.) and create an archive (on Windows a zip file, on Unix a tar file). You
//! should be able to distribute this archive with no problems; if there are any, please open an
//! issue.
//!
//! ### Deploying for the web
//!
//! If you're deploying for the web, first make sure you've
//! [installed the cargo web tool](https://github.com/koute/cargo-web). Then use `cargo web deploy`
//! to build your application for distribution (located at `target/deploy`).
//!
//! If you want to test your application locally, use `cargo web start --features quicksilver/stdweb` and open your
//! favorite browser to the port it provides.
//!
//! #### wasm-bindgen support
//!
//! Quicksilver has recently gained experimental support for `wasm-bindgen`, under the `web-sys`
//! feature. The workflow is not currently documented here, but it should be the same as using any other
//! library with `wasm-bindgen`.
//!
//! ## Optional Features
//!
//! Quicksilver by default tries to provide all features a 2D application may need, but not all
//! applications need these features.
//!
//! The optional features available are:
//! - easy logging (via [log](https://github.com/rust-lang/log),
//! [simple_logger](https://github.com/borntyping/rust-simple_logger), and
//! [web_logger](https://github.com/yewstack/web_logger))
//! - gamepad event generation (via [gilrs](https://gitlab.com/gilrs-project/gilrs))
//! - saving (via [gestalt](https://github.com/ryanisaacg/golem))
//! - font rendering (via [elefont](https://github.com/ryanisaacg/elefont)) and TTF parsing (via [rusttype](https://gitlab.redox-os.org/redox-os/rusttype))
//!
//! Each are enabled by default, but you can
//! [specify which features](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#choosing-features)
//! you actually want to use.
//!
//! ## Supported Platforms
//!
//! The engine is supported on Windows, macOS, Linux, and the web via WebAssembly.
//!
//! Mobile support would be a future possibility, but likely only through external contributions.

#![deny(
    bare_trait_objects,
    unused_extern_crates,
    unused_import_braces,
    unused_qualifications
)]
#![allow(clippy::needless_doctest_main)]

// Re-export every library that appears in the public API
pub use blinds;
#[cfg(feature = "font")]
pub use elefont;
pub use golem;
pub use log;
pub use mint;

mod error;

pub mod geom;
pub mod graphics;
pub mod input;
#[cfg(feature = "saving")]
pub mod saving {
    //! A module to manage cross-platform save data via the [`gestalt`] library
    pub use gestalt::*;
}
pub use crate::error::QuicksilverError;

mod run;
mod timer;
mod window;
pub use blinds::CursorIcon;
pub use run::{run, Settings};
pub use timer::Timer;
pub use window::Window;

pub use graphics::Graphics;
pub use input::Input;

/// Load a file as a [`Future`]
///
/// Within an `async` function (like the one passed to [`run`]), you can use `.await`:
///
/// ```no_run
/// # use platter::load_file;
/// # async fn test() {
/// load_file("my_file_path").await.expect("The file was not found!");
/// # }
/// ```
///
/// [`Future`]: std::future::Future
/// [`run`]: crate::ran
pub use platter::load_file;

/// A Result that returns either success or a [`QuicksilverError`]
pub type Result<T> = std::result::Result<T, QuicksilverError>;

// Evil hack to preserve docs from Blinds
// The events from Blinds want to link to crate::event::*,
// so this creates those links
mod event {
    pub use crate::input::*;
}
